lms/djangoapps/instructor_task/tasks_helper/grades.py

```diff
@@ -297,9 +297,26 @@ def grouper(iterable, chunk_size=self.USER_BATCH_SIZE, fillvalue=None):
             args = [iter(iterable)] * chunk_size
             return izip_longest(*args, fillvalue=fillvalue)
 
-        users = CourseEnrollment.objects.users_enrolled_in(context.course_id, include_inactive=True)
-        users = users.select_related('profile')
-        return grouper(users)
+        def users_for_course(course_id):
```

Comment:
@fysheets `def _batch_users(self, context):` is basically a generator and its body is only executed during the iteration of the `generator object` (i.e. like here: https://github.com/edx/edx-platform/blob/0afc7a7ec86071258187e78d81d6147221499379/lms/djangoapps/instructor_task/tasks_helper/grades.py#L249). The both for loops are necessary. 

The third loop which you mentioned `for user, course_grade, error in CourseGradeFactory().iter(......)` is just a class method to named `iter` here. 
https://github.com/edx/edx-platform/blob/0afc7a7ec86071258187e78d81d6147221499379/lms/djangoapps/instructor_task/tasks_helper/grades.py#L450-L455


https://github.com/edx/edx-platform/blob/5289c7e828ddca1ac9a745bb1a84e0133d3ce77d/lms/djangoapps/grades/course_grade_factory.py#L79-L100


I looked at this today, and I couldn't find any good workaround to avoid any of those two loops. Let me know what you think or if you have any idea let me know. ?